### ISSUE TITLE
Data API - Supported operations

### ISSUE THREAD

**Issue by raynaudoe** [Zondax/rosetta-filecoin-proxy#6]
Rosetta describes an operation as the struct:
```
type Operation struct {
	OperationIdentifier *OperationIdentifier `json:"operation_identifier"`
	RelatedOperations []*OperationIdentifier `json:"related_operations,omitempty"`
	Type string `json:"type"`
	Status   string                 `json:"status"`
	Account  *AccountIdentifier     `json:"account,omitempty"`
	Amount   *Amount                `json:"amount,omitempty"`
	Metadata map[string]interface{} `json:"metadata,omitempty"`
}
```
```

// OperationIdentifier The operation_identifier uniquely identifies an operation within a
// transaction.
type OperationIdentifier struct {
	Index int64 `json:"index"`
	NetworkIndex *int64 `json:"network_index,omitempty"`
}
```

```
// AccountIdentifier The account_identifier uniquely identifies an account within a network. All
// fields in the account_identifier are utilized to determine this uniqueness (including the
// metadata field, if populated).
type AccountIdentifier struct {
	Address    string                `json:"address"`
	SubAccount *SubAccountIdentifier `json:"sub_account,omitempty"`
	Metadata map[string]interface{} `json:"metadata,omitempty"`
}
```

```
// Amount Amount is some Value of a Currency. It is considered invalid to specify a Value without a
// Currency.
type Amount struct {
	// Value of the transaction in atomic units represented as an arbitrary-sized signed integer.
	// For example, 1 BTC would be represented by a value of 100000000.
	Value    string                 `json:"value"`
	Currency *Currency              `json:"currency"`
	Metadata map[string]interface{} `json:"metadata,omitempty"`
}
```

```
// Currency Currency is composed of a canonical Symbol and Decimals. This Decimals value is used to
// convert an Amount.Value from atomic units (Satoshis) to standard units (Bitcoins).
type Currency struct {
	Symbol string `json:"symbol"`
	Decimals int32 `json:"decimals"`
	Metadata map[string]interface{} `json:"metadata,omitempty"`
}
```

Must check values mappings against FileCoin's.

**Comment by raynaudoe** [comment]
On Lotus api, in order to get method's string pointed by `Message.Method`, first you get the actor from the source address:
 `actor, _ := api.StateGetActor(context.Background(), msg.Message.From, filTypes.EmptyTSK)`
Then by using `actor.Code` we can switch between the available methods that match to the actor's code.

```
switch actor.Code {
	case builtin.InitActorCodeID:
		method = builtin.MethodsInit
	case builtin.CronActorCodeID:
		method = builtin.MethodsCron
	case builtin.AccountActorCodeID:
		method = builtin.MethodsMultisig
...
}
```
The main problem is that calling `api.StateGetActor` for each message on a block, makes the whole endpoint call so slow that rosetta-cli errors with time out.

Maybe there's another way on FileCoin's api to get actor's method string ?

**Comment by qiwu7** [comment]
Found an issue that all the **value** field of a operation is reported as 0.

sample request
```
curl --location --request POST '{{NODE_URL}}/block' \
--header 'Content-Type: application/json' \
--data-raw '{
    "network_identifier": {
        "blockchain": "Filecoin",
        "network": "testnet"
    },
    "block_identifier": {
        "index": 139362      
    }
}'
```

response
<details>
  <summary>Click to expand!</summary>
  
```
{
    "block": {
        "block_identifier": {
            "index": 139362,
            "hash": "0171a0e402207f01511a97868bf8663711b4b276884517acb590446c7f777520b5fae5e54473"
        },
        "parent_block_identifier": {
            "index": 139359,
            "hash": "0171a0e4022036824e1761b6b9a25571e77fc1a19f3434aff65870e00f5ca935df3fd8e291ef"
        },
        "timestamp": 1596008850000,
        "transactions": [
            {
                "transaction_identifier": {
                    "hash": "bafy2bzacec56j6s4lojh4vefuxzhf4al6p2jks63jwtkdp6zz6kl67hntv5hs"
                },
                "operations": [
                    {
                        "operation_identifier": {
                            "index": 0
                        },
                        "type": "Transfer",
                        "status": "Ok",
                        "account": {
                            "address": "t3rbq4ijj5owypedib5txu7fbmeq3x27znjgd3ctpchmuu27gilaqnmqe5vw3v6emtv7tzj3trjbhpux4j4m4a"
                        },
                        "amount": {
                            "value": "0",
                            "currency": {
                                "symbol": "FIL",
                                "decimals": 18
                            }
                        }
                    },
                    {
                        "operation_identifier": {
                            "index": 1
                        },
                        "related_operations": [
                            {
                                "index": 0
                            }
                        ],
                        "type": "Transfer",
                        "status": "Ok",
                        "account": {
                            "address": "t04067"
                        },
                        "amount": {
                            "value": "0",
                            "currency": {
                                "symbol": "FIL",
                                "decimals": 18
                            }
                        }
                    }
                ]
            },
            {
                "transaction_identifier": {
                    "hash": "bafy2bzacecdayy3niq4kjukg3shrq55fcnmhmfxnipxb7wa2eakd6xjgggyde"
                },
                "operations": [
                    {
                        "operation_identifier": {
                            "index": 0
                        },
                        "type": "Transfer",
                        "status": "Ok",
                        "account": {
                            "address": "t3rbq4ijj5owypedib5txu7fbmeq3x27znjgd3ctpchmuu27gilaqnmqe5vw3v6emtv7tzj3trjbhpux4j4m4a"
                        },
                        "amount": {
                            "value": "0",
                            "currency": {
                                "symbol": "FIL",
                                "decimals": 18
                            }
                        }
                    },
                    {
                        "operation_identifier": {
                            "index": 1
                        },
                        "related_operations": [
                            {
                                "index": 0
                            }
                        ],
                        "type": "Transfer",
                        "status": "Ok",
                        "account": {
                            "address": "t04067"
                        },
                        "amount": {
                            "value": "0",
                            "currency": {
                                "symbol": "FIL",
                                "decimals": 18
                            }
                        }
                    }
                ]
            }
        ],
        "metadata": {
            "blockCIDs": [
                "bafy2bzacebbejv7zexyxlg3hltx5xo6n6iu6thmh6fnrq3indysrm3zk63fgk"
            ]
        }
    }
}
```
</details>

**Comment by raynaudoe** [comment]
Maybe those transactions weren't "Transfers" at all (until we find a way to solve method's name, all operation are labeled as "Transfer"), thus the `0` amount...
Also, the `amount` is directly taken from the `Message` itself, here:
https://github.com/Zondax/rosetta-filecoin-proxy/blob/d8c58100ea64764edb462c8a87eece26fd589139/rosetta/services/block.go#L159

**Comment by qiwu7** [comment]
Ok this is the [transaction](https://filfox.info/en/message/bafy2bzaceahh24s4xdy5uruarjtpbx26g2ebpngrzn7qfp6v3ymqmgowlocwo?block=bafy2bzacebbejv7zexyxlg3hltx5xo6n6iu6thmh6fnrq3indysrm3zk63fgk) I am looking for, and I realized that it's not in the response, and it is a send. ideas?

**Comment by raynaudoe** [comment]
Ok, this is what's happening:
the `/block` endpoint will return only the messages that were _executed_ on that block. Messages _generated_ at height `N` will be _executed_ on any block at height `N+1`. 
On the particular case you mention, the transactions you see are the ones that were generated at height  `139361`. If you query `/block` at height `139363` you'll find the transaction.

If you need this behavior to be changed, please open a new issue.

**Comment by qiwu7** [comment]
Yes it makes sense. I almost forgot that the messages are executed in the N+1 block.

Another interesting that related to this is when you implement `/block/transaction` endpoint, it would be very tricky since we don't know we want to show the message in the current block or the executed message in the next block...

**Comment by raynaudoe** [comment]
Coming back to the original problem on getting method's name, according to Filecoin's team, there is not a fast or easy way to get this. 
Highly probable that changes inside Lotus are needed to provide a faster/easier way on getting method's name.

**Comment by whyrusleeping** [comment]
To get the method names, theres just a static table of all method IDs here: https://github.com/filecoin-project/specs-actors/blob/master/actors/builtin/methods.go

> The main problem is that calling api.StateGetActor for each message on a block, makes the whole endpoint call so slow that rosetta-cli errors with time out.

This is very easily cacheable, you will only need to call StateGetActor once per actor. The call should be simple to memoize clientside. At one call per actor, you're looking at ~16,000 calls on the current space race network (which admittedly is young) at about 30ms per call (this number may be a bit high, i'm running a miner on the same box i timed this on). 480 seconds in total, absolute maximum of around 18 seconds per block if every message in the new block is to a new actor, and none of them are cached. These requests can also be done in parallel, and you can use some tricks to short circuit these things as well, for example, any t1 or t3 address is always an account actor, and the 'builtin' actors can be hard coded (they are also some of the most frequently interacted with actors in the system).

**Comment by jbenet** [comment]
Hey @raynaudoe in [this comment](https://github.com/Zondax/rosetta-filecoin-proxy/issues/6#issuecomment-665207218), 
> On Lotus api, in order to get method's string pointed by `Message.Method`, first you get the actor from the source address: 
>`actor, _ := api.StateGetActor(context.Background(), msg.Message.From, filTypes.EmptyTSK)`

I'm not sure, as idk how you're really using it or what the whole code looks like, but at a glance, I _think_ that if you want the actor that `msg.Message.Method` is bound to, you want `msg.Message.To`-- the _destination_ address:

> `actor, _ := api.StateGetActor(context.Background(), msg.Message.To, filTypes.EmptyTSK)`

Invocations in the Filecoin "VM" should be on the destination actor -- conceptually: `call(m.To, m.Method, m.Params)` or said another way: `to.method(params...)`

**Comment by raynaudoe** [comment]
Ok! Thanks both @whyrusleeping @jbenet !!

I'll try implementing some sort of cache and see how it goes

**Comment by jleni** [comment]
@qiwu7 I think it would still be useful if rosetta would a) increase the timeout or better.. b) consider a few retries. 

This is just to cover extreme cases where some tipset has a large number of new accounts/actors and, even with caching in place, it is not possible to respond in time. Very likely on the second try, caching will speed things up and the response will be available on time.

**Comment by jleni** [comment]
I am talking about both. 
I think retries if there is a timeout when fetching a block should be enough.

The fact is.. if you call the rosetta-proxy `/block` endpoint, normally the cache allow answering quickly, but there will be some very unlikely cases, where the call will take >30s. Internally, the proxy will be collecting data and filling up the cache.. so a second call will be serviced quickly.

**Comment by raynaudoe** [comment]
Hi @qiwu7 , we've a PR almost ready for using the new actors db that will return the actual method's name on each transaction. 
The only thing missing is to define which Filecoin's operations will CB support, so I can add them on rosetta's supported list.

**Comment by qiwu7** [comment]
The quick answer is types for all the balance change events. I have listed out the potential balance changing events that I image could affect balance changing in an earlier thread.

- Transfer
- Fee
- Reward
- Fund Unlock (Transfer?)
- SwapKey?

**Comment by raynaudoe** [comment]
Ok done it, thanks @qiwu7.

Should the endpoints `/mempool/` and `/mempool/transaction`  only show supported operations ? Or show all transactions types ?

**Comment by qiwu7** [comment]
we are not leveraging `/mempool` at the moment. But I think it makes sense to show all types in the mempool.

**Comment by qiwu7** [comment]
@raynaudoe I am not super familiar with all the types. So I have only listed out the balance changing events that are necessary right now. But Feel free to include other types of operations if that makes sense. For instance, governance related (voting) or something similar, which could also be useful.

**Comment by jleni** [comment]
I suggest getting the more basic right now and release 
We can open another issue and PR for the more esoteric ones later on

**Comment by qiwu7** [comment]
yeah that works as long as we are tacking it :)

**Comment by qiwu7** [comment]
Can we add some comments/clarifications on what these methods are?
```
    "Send":             true, //Common
	"AwardBlockReward": true, //MethodsReward
	"ThisEpochReward":  true, //MethodsReward
	"SwapSigner":       true, //MethodsMultisig
	"LockBalance":      true, //MethodsMultisig
	"AddBalance":       true, //MethodsMarket
	"WithdrawBalance":  true, //MethodsMarket
```
For example
- `AwardBlockReward`. Is this the block reward to miner?
- `ThisEpochReward`, what’s difference between this and `AwardBlockReward`
-  `LockBalance`, is unlocking the funds or locking the funds?
- `Send`, does this include a multisig send/transfer? Maybe we rename this to Payment  or Transfer? Since the operation is actually a positive amount on the receiver and negative amount on the sender. It looks wired if this is a SEND with negative amount on that operation. So might be better if we name it a payment/transfer op
What does the AddBalance and WithdrawBalance  for the market means? I don’t think really pertains to us right but it would be great to add some more comments :slightly_smiling_face:
- `FEE`, is this missing? I don’t see a fee operation in the list

**Comment by raynaudoe** [comment]
Closing issue, moving discussion to #66
